### ISSUE TITLE
InputSystemDisable() followed by InputSystem.Enable causes NullPointerExceptions

### ISSUE THREAD

**Issue by julenka** [microsoft/MixedRealityToolkit-Unity#5085]
## Overview
Disabling and then re-enabling the input system causes the following exception to be thrown every frame:

```
NullReferenceException: Object reference not set to an instance of an object
Microsoft.MixedReality.Toolkit.Input.FocusProvider.RegisterPointers (Microsoft.MixedReality.Toolkit.Input.IMixedRealityInputSource inputSource) (at Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs:689)
````

## Steps to reproduce
Use the following script to enable and disable the input system. Notice that after enabling the input system console spits out continuous null pointers:

```
using Microsoft.MixedReality.Toolkit;
using Microsoft.MixedReality.Toolkit.Input;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class DisableInputSystemTest : MonoBehaviour
{
    private IMixedRealityInputSystem inputSystem = null;
    private IMixedRealityInputSystem InputSystem
    {
        get
        {
            if (inputSystem == null)
            {
                MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem);
            }
            return inputSystem;
        }
    }

    public void DisableInputSystem()
    {
        InputSystem.Disable();
    }

    public void EnableInputSystem()
    {
        InputSystem.Enable();
    }

    public void Update()
    {
        if (Input.GetKeyDown(KeyCode.Alpha1))
        {
            EnableInputSystem();
        }
        if (Input.GetKeyDown(KeyCode.Alpha2))
        {
            DisableInputSystem();
        }
    }
}
```

## Unity Editor Version
2018.4.1f1

## Mixed Reality Toolkit Release Version
mrtk_development latest

**Comment by julenka** [comment]
I think I have a solution to this, described in https://stackoverflow.com/questions/56757620/how-to-temporarly-disable-mixedrealitytoolkit-inputsystem/56783236#56783236

**Comment by gilbdev** [comment]
Hey Julia, Thanks for helping,
AS said, when using Enable after Disable, i got a lot of errors on mrtk_master RC2.1
here are the complete callstacks i can't put in a comment on stackoverflow..

first time:
```
ArgumentNullException: Value cannot be null.
Parameter name: _unity_self
Microsoft.MixedReality.Toolkit.Windows.Input.GestureRecognizerExtensions.UpdateAndResetGestures (UnityEngine.XR.WSA.Input.GestureRecognizer recognizer, UnityEngine.XR.WSA.Input.GestureSettings gestureSettings) (at Assets/MixedRealityToolkit/Extensions/GestureRecognizerExtensions.cs:11)
Microsoft.MixedReality.Toolkit.WindowsMixedReality.Input.WindowsMixedRealityDeviceManager.set_GestureSettings (Microsoft.MixedReality.Toolkit.Windows.Input.WindowsGestureSettings value) (at Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityDeviceManager.cs:178)
Microsoft.MixedReality.Toolkit.WindowsMixedReality.Input.WindowsMixedRealityDeviceManager.Enable () (at Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityDeviceManager.cs:271)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterServiceInternal (System.Type interfaceType, Microsoft.MixedReality.Toolkit.IMixedRealityService serviceInstance) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:834)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterServiceInternal[T] (T serviceInstance) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:849)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterService[T] (System.Type concreteType, Microsoft.MixedReality.Toolkit.Utilities.SupportedPlatforms supportedPlatforms, System.Object[] args) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:206)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterDataProvider[T] (System.Type concreteType, Microsoft.MixedReality.Toolkit.Utilities.SupportedPlatforms supportedPlatforms, System.Object[] args) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:301)
Microsoft.MixedReality.Toolkit.Input.MixedRealityInputSystem.Enable () (at Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs:221)
```
second time, got 2 different errors:
```
There's already a IMixedRealityInputDeviceManager.Windows Mixed Reality Device Manager registered!
UnityEngine.Debug:LogError(Object)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterServiceInternal(Type, IMixedRealityService) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:812)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterServiceInternal(IMixedRealityInputDeviceManager) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:849)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterService(Type, SupportedPlatforms, Object[]) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:206)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterDataProvider(Type, SupportedPlatforms, Object[]) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:301)
Microsoft.MixedReality.Toolkit.Input.MixedRealityInputSystem:Enable() (at Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs:221)
```

```
ArgumentException: An item with the same key has already been added. Key: Flat
System.Collections.Generic.Dictionary`2[TKey,TValue].TryInsert (TKey key, TValue value, System.Collections.Generic.InsertionBehavior behavior) (at <d7ac571ca2d04b2f981d0d886fa067cf>:0)
System.Collections.Generic.Dictionary`2[TKey,TValue].Add (TKey key, TValue value) (at <d7ac571ca2d04b2f981d0d886fa067cf>:0)
Microsoft.MixedReality.Toolkit.Utilities.ArticulatedHandPose.LoadGesturePose (Microsoft.MixedReality.Toolkit.Utilities.ArticulatedHandPose+GestureId gesture, System.String filePath) (at Assets/MixedRealityToolkit/Definitions/Utilities/ArticulatedHandPose.cs:218)
Microsoft.MixedReality.Toolkit.Utilities.ArticulatedHandPose.LoadGesturePoses () (at Assets/MixedRealityToolkit/Definitions/Utilities/ArticulatedHandPose.cs:208)
Microsoft.MixedReality.Toolkit.Input.InputSimulationService.Initialize () (at Assets/MixedRealityToolkit.Services/InputSimulation/Editor/InputSimulationService.cs:65)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterServiceInternal (System.Type interfaceType, Microsoft.MixedReality.Toolkit.IMixedRealityService serviceInstance) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:833)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterServiceInternal[T] (T serviceInstance) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:849)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterService[T] (System.Type concreteType, Microsoft.MixedReality.Toolkit.Utilities.SupportedPlatforms supportedPlatforms, System.Object[] args) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:206)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.RegisterDataProvider[T] (System.Type concreteType, Microsoft.MixedReality.Toolkit.Utilities.SupportedPlatforms supportedPlatforms, System.Object[] args) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:301)
Microsoft.MixedReality.Toolkit.Input.MixedRealityInputSystem.Enable () (at Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs:221)
```

following times, got 2 errors each time, 
the There's already a IMixedRealityInputDeviceManager.Windows Mixed Reality Device Manager registered!`` above +
```
There's already a IMixedRealityInputDeviceManager.Input Simulation Service registered!
UnityEngine.Debug:LogError(Object)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterServiceInternal(Type, IMixedRealityService) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:812)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterServiceInternal(IMixedRealityInputDeviceManager) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:849)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterService(Type, SupportedPlatforms, Object[]) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:206)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:RegisterDataProvider(Type, SupportedPlatforms, Object[]) (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:301)
Microsoft.MixedReality.Toolkit.Input.MixedRealityInputSystem:Enable() (at Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs:221)
FramaToolkit.<DoLoad>d__23:MoveNext() (at Assets/FramaToolkit.Core/Loading/SceneLoader.cs:134)
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
```

scenario is disabling input while loading additive scenes, this is made in a coroutine (it may count)
simplified coroutine (without loading scene stuff):
```
private IEnumerator DoLoad()
		{
			//load loading scene
			Debug.Log($"[Loading] start");

			InputSystem.Disable();

			//start loading timer
			float startLoadingTime = Time.realtimeSinceStartup;

			//wait the minimum time
			yield return new WaitUntil(() => MinLoadTimeSec <= (Time.realtimeSinceStartup - startLoadingTime));

			InputSystem.Enable();

			Debug.Log($"[Loading] end");

		}
```

**Comment by gilbdev** [comment]
trying in mrtk_development + julia suggested changes

this is almost working as expected but i still got an error : 


```
            // Update all registered runtime services
l.954:    foreach (var service in registeredMixedRealityServices)
            {
                service.Item2.LateUpdate();
            }
```

```
InvalidOperationException: Collection was modified; enumeration operation may not execute.
System.ThrowHelper.ThrowInvalidOperationException (System.ExceptionResource resource) (at <d7ac571ca2d04b2f981d0d886fa067cf>:0)
System.Collections.Generic.List`1+Enumerator[T].MoveNextRare () (at <d7ac571ca2d04b2f981d0d886fa067cf>:0)
System.Collections.Generic.List`1+Enumerator[T].MoveNext () (at <d7ac571ca2d04b2f981d0d886fa067cf>:0)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.LateUpdateAllServices () (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:954)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit.LateUpdate () (at Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs:665)
```
=> the disable service should be push in a pending change list, and postponed at the end of the frame. apparently here i have trigger the disable during another lateupdate of a service (i was just clicking on a button :( )
should it be handled in mrtk directly ?

**Comment by julenka** [comment]
Maybe Kurtis could help here?

**Comment by gilbdev** [comment]
on a longer term : 

if you want users to avoid accessing some functions, you may go with internal interfaces.

http://www.csharp411.com/c-internal-interface/

together with  [InternalsVisibleToAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute) to be able to have internal calls cross-mrtk-assemblies

```
internal interface IMixedRealityServiceInternal
{
	void MyInternalMethod();
}
public class MyService :
	IMixedRealityService, IMixedRealityServiceInternal
{
	public void MyPublicMethod() { }
	void IMixedRealityServiceInternal.MyInternalMethod()
	{
	}
}
```
the only tricky thing is you have to cast (or store) the MyService to IMixedRealityServiceInternal to be able to call the internal function or use [some extensions tricks](https://alexfranchuk.com/blog/internal-interface-classes-in-csharp/)

this may be a lot of refactor but will clearly show what can be called from the public interface to users :)

**Comment by julenka** [comment]
>Document that the IMixedRealityService functions aren't meant to be called except by the MRTK runtime itself.
> Update the IMixedRealityServiceRegistrar to have an EnableService and DisableService function, which will properly enable/disable the corresponding services (for example, in the input system case, it will also disable/enable the focus provider gaze provider)
> Add a helper to CoreServices to also have this EnableService/DisableService to help with familiarity for folks who are already using the CoreServices helper.
> Move the registration and management of the focusprovider (and related services) inside of the input system.

These make sense to me.

> I am not 100% certain why focusprovider is considered a core system, though i can imagine scenarios where it would be handy to be separate from the input system.

I also don't see why. I guess that Stephen H. the original architect could maybe explain why he made FocusProvider a core system? I would think that FocusProvider is just an internal part of the input system. I don't really see how focus provider could be used independently of the input system. Seems like we should probably remove it, though that's probably not as important as issues right now (fixing functionality, adding documentation).

**Comment by connecttoconnecticut** [comment]
+1 with this error. 

Interesting thing is that, i have two hololenses 1. I have problem on one when i return to my app, lets say, get focus back on app, my hololens does not recognize voice commands any more, but on second, it works perfectly clear. And yes, its same version of OS, last updates are done.

I am not 100% sure, but i guess this bug is in relation with this.

And now i came to this error with trying to fix this, with reloading scene onappfocus, but here i am, with this same error.

**Comment by wiwei** [comment]
@gilbdev FYI, I'm not sure that I'm going to be able to push in my change for the current iteration because I'm going to be out for a little bit and won't be able to get to the migration tooling before then. Because my change is breaking (in that a thing that is a MonoBehaviour is no longer one), I think that providing the migration tooling is SUPER important before going out.

That being said, I wanted to send out the change so that folks who were interested in grabbing that functionality could unblock themselves - the overall shape of this *shouldn't* change that much.

**Comment by wiwei** [comment]
@gilbdev, can you share a little more on your scenario in which you're using the Reset() or Disable/Enable functionality? Some folks have asked some questions regarding the right way of addressing this and I want to make sure that the linked PR actually does address the scenario you're trying to accomplish.

**Comment by gilbdev** [comment]
overall scenario : i have a loading sequence to switch between scenes during which i fade out and in.
during the loading sequence, i want to avoid any input  on buttons, voice commands or any other kind, hence trying to disable the whole input system.

- load initial root scene with managers and mrtk
- launch loading sequence
-disable input
- fade
- (opt.) unload previous scenes (except root ^^)
- load new scenes
-wait at least 1sec for whole loading sequence
- fade
- enable back input
- play and redo from step 2 as necessary

**Comment by wiwei** [comment]
@gilbdev do you want to be able to poll for new input state during that transition (i.e. if you were to ask InputSystem.GazeProvider.GetTheHeadPosition or InputSystem.WhereAreTheHands) would you expect those things to work, or would you expect for the input system to be completely shutdown during that time (or would you not care, as long as the net effect is the same?)

**Comment by gilbdev** [comment]
complete shutdown.
 i could have some poll (like voice commands i can't mute the player ^^, or even a button that is still not unloaded, or already laoded but not visible due to fade not completed) but it should have zero effect. i think the focuspointer should not be visible either (that's something we can do on our side if it's not automatic by the fact of disabling the system)
and when enabling back, all goes to 'normal' (for example if i focus on a button, it should switch to focus state when system turns enabled -- but this may be link to the whole intereactable refactor too at this point, just giving it as example)

**Comment by wiwei** [comment]
Alright, thanks for clarifying that!
